File: bindings/nodejs/cli-native.mjs
#!/usr/bin/env node

import path from 'node:path';
import fs from 'node:fs';
import { parse } from 'csv-parse/sync';

/**
 * A simple CLI to calculate statistics from a CSV file using a native Node.js
 * implementation.
 */
function main() {
  // Basic command-line argument parsing.
  // process.argv[0] is the node executable
  // process.argv[1] is the path to this script
  const filePath = process.argv[2];

  if (!filePath) {
    console.error('Usage: ./cli-native.mjs <path-to-csv-file>');
    process.exit(1);
  }

  // Resolve the path to an absolute path for clarity.
  const absolutePath = path.resolve(filePath);
  console.log(`Calculating statistics for '${absolutePath}'...`);

  try {
    // Call the core logic from our native implementation.
    const stats = calculateStatsFromFile(absolutePath);

    // If successful, print the statistics in a formatted way.
    console.log("\n--- Statistics for 'Amount Received' ---");
    console.log(`Count:              ${stats.count}`);
    console.log(`Min:                ${stats.min.toFixed(4)}`);
    console.log(`Max:                ${stats.max.toFixed(4)}`);
    console.log(`Sum:                ${stats.sum.toFixed(4)}`);
    console.log(`Mean:               ${stats.mean.toFixed(4)}`);
    console.log('----------------------------------------');
  } catch (e) {
    // Print the error message to stderr and exit.
    console.error(`Error: ${e.message}`);
    process.exit(1);
  }
}

function calculateStatsFromFile(filePath) {
  const fileContent = fs.readFileSync(filePath, { encoding: 'utf8' });

  // Use csv-parse to process the file.
  const records = parse(fileContent, {
    columns: true, // Treat the first line as headers.
    skip_empty_lines: true // Ignore empty lines.
  });

  let count = 0;
  let min = Infinity;
  let max = -Infinity;
  let sum = 0;

  // Process each record using a for...of loop.
  for (const record of records) {
    const amountStr = record['Amount Received'];

    if (amountStr) {
      const amount = parseFloat(amountStr);
      // Ensure the parsed amount is a valid number.
      if (!isNaN(amount)) {
        count++;
        sum += amount;
        if (amount < min) min = amount;
        if (amount > max) max = amount;
      }
    }
  }

  // Avoid division by zero if there are no valid data points.
  const mean = count > 0 ? sum / count : 0;

  // If no data was found, min/max should be 0.
  if (count === 0) {
    min = 0;
    max = 0;
  }

  return { count, min, max, sum, mean };
}

// Run the main function.
main();

File: bindings/python/python/cli-native.py
#!/usr/bin/env python

import argparse
import sys
import csv

class Statistics:
    """A simple data class to hold the calculated statistics."""
    def __init__(self, count, min_val, max_val, total_sum, mean):
        self.count = count
        self.min = min_val
        self.max = max_val
        self.sum = total_sum
        self.mean = mean

def calculate_stats_from_file(file_path):
    """
    Reads a CSV file and calculates statistics for the "Amount Received" column.

    This implementation uses pure Python for performance comparison.

    Args:
        file_path (str): The path to the CSV file.

    Returns:
        Statistics: An object containing the count, min, max, sum, and mean.

    Raises:
        ValueError: If the file is not found, cannot be processed, is empty,
                    or does not contain the required column.
    """
    count = 0
    total_sum = 0.0
    min_val = float('inf')
    max_val = float('-inf')

    try:
        with open(file_path, mode='r', encoding='utf-8') as infile:
            reader = csv.DictReader(infile)

            if "Amount Received" not in reader.fieldnames:
                raise ValueError("CSV file must have a column named 'Amount Received'")

            for row in reader:
                try:
                    # Extract and convert the amount, skipping if invalid
                    amount = float(row["Amount Received"])
                    count += 1
                    total_sum += amount
                    if amount < min_val:
                        min_val = amount
                    if amount > max_val:
                        max_val = amount
                except (ValueError, TypeError):
                    # This handles rows with non-numeric data or empty values
                    continue
    except FileNotFoundError:
        raise ValueError(f"Error: File not found at '{file_path}'")
    except Exception as e:
        raise ValueError(f"Error processing file: {e}")

    if count == 0:
        raise ValueError("No valid numeric data found in 'Amount Received' column.")

    mean = total_sum / count

    return Statistics(
        count=count,
        min_val=min_val,
        max_val=max_val,
        total_sum=total_sum,
        mean=mean
    )

def main():
    """
    A simple CLI to calculate statistics from a CSV file using pure Python.
    This script serves as a baseline for performance comparison against other
    implementations (e.g., a Rust-powered one).
    """
    parser = argparse.ArgumentParser(
        description="Calculate statistics from a CSV file using pure Python."
    )
    parser.add_argument(
        "file_path",
        type=str,
        help="The path to the CSV file to process.",
    )

    args = parser.parse_args()
    file_path = args.file_path

    print(f"Calculating statistics for '{file_path}' (Native Python)...")

    try:
        # Call the pure Python implementation for statistics calculation.
        stats = calculate_stats_from_file(file_path)

        # If successful, print the statistics in a formatted way.
        print("\nOutput for python-native")
        print("--- Statistics for 'Amount Received' ---")
        print(f"Count:              {stats.count}")
        print(f"Min:                {stats.min:.4f}")
        print(f"Max:                {stats.max:.4f}")
        print(f"Sum:                {stats.sum:.4f}")
        print(f"Mean:               {stats.mean:.4f}")
        print("----------------------------------------")

    except ValueError as e:
        # The calculation function raises ValueError for known issues.
        # Print the error message to stderr and exit.
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()
